### ISSUE TITLE
Windows users - problem installing vagrant plugins?

### ISSUE THREAD

**Issue by mattkoch614** [ctsit/redcap_deployment#34]
Good morning,

We're very interested in trying this out as a local development environment for REDCap, but are having a small issue - we can't seem to get the two required Vagrant plugins installed.

```
vagrant plugin install vagrant-hostsupdater
vagrant plugin install vagrant-env
```

```
C:/HashiCorp/Vagrant/embedded/mingw64/lib/ruby/2.3.0/resolv.rb:766:in `recvfrom': The connection 
has been broken due to keep-alive activity detecting a failure while the operation was in progress. - 
recvfrom(2) (Errno::ENETRESET)
```
Do any other Windows users have this problem? 

It may be due to our Firewall settings, but I've been unable to nail it down.  I realize this isn't an issue specifically related to this project, but I figured I'd post it here in case others were having similar problems.

**Comment by pbchase** [comment]
Matt, I can't speak to the vagrant problem, I can address the matter of running our tools on Windows. To the best of my knowledge, redcap_deployment has never been tested on Windows. We wrote on Mac OSX and use it every day on Mac OSX.  We expect it would work on any Mac with GNU Utilities installed and any Linux host.  

The Fabric tasks we wrote calls lots of local commands I expect would run only on a POSIX host like Linux or Mac OSX.  These are the local commands we call: bash, cd, cp, find, git, ls, mkdir, mktemp, python, rm, test, unzip. 

Sorry, but I think the vagrant issue will be the least of your problems.

**Comment by mattkoch614** [comment]
@pbchase Thanks for the reply! Unless I'm mistaken, I think installing Git for Windows (e.g. Git Bash) would probably provide enough of a POSIX environment to support the commands you listed, would it not? Most/all of us already have this installed in order to pull down repos.

https://git-scm.com/download/win

Since this hasn't been fully tested on Windows, I think it would be an interesting experiment to try and get it up and running. If we manage to do that, perhaps I can write up a little how-to on what we ended up doing and post it somewhere.

**Comment by pbchase** [comment]
@mattkoch614, yes, git-bash might well have all the support you need. We would certainly welcome instructions or code modifications to enable Windows support. Please send us your pull requests. 
We're not in a position to test them on Windows, but we could certainly test them on Mac OSX to make sure there are no regressions there. 

FYI, I released 2.0.0 of redcap_deployment tonight. This release is significant in that it drops support for hooks entirely. Our strategic direction for REDCap extensions is external modules. This has allowed us to drop the last vestiges of a hook architecture that was probably unique to UF. I encourage you test and develop on 2.0.0 or higher.

**Comment by mattkoch614** [comment]
A couple of updates from my end:

- Disabling Symantec Endpoint Protection allowed us to install the vagrant plugins without issue.
- Downloaded a REDCap installation (.zip) file from the REDCap Community site and dropped it into the repository root.
- After installing the vagrant plugins and creating a ``.env`` file with the recommended settings, the ``vagrant up`` command kicks off successfully.
- We ran into a problem during the ``bootstrap.sh`` script. A snippet of the error is below:

```
==> default: Running provisioner: shell...
    default: Running: C:/Users/kochm/AppData/Local/Temp/vagrant-shell20180129-10140-cifp6c.sh
==> default: Import environment variables from /vagrant/.env
==> default: /vagrant/.env: line 11: $'\r': command not found
The SSH command responded with a non-zero exit status. Vagrant
assumes that this means the command failed. The output for this command
should be in the log above. Please read the output to determine what
went wrong.
```

**Comment by pbchase** [comment]
I'm 99% sure you're right about that. BASH does _not_ like CRLF line endings. This article provides a few ways to address the issues within git: [Mind the End of Your Line](http://adaptivepatchwork.com/2012/03/01/mind-the-end-of-your-line/).  I recommend you add a .gitattributes file to the root of the repo with the content

`*.sh crlf=input`
`.env crlf=input`

You'll probably need to checkout the affected files again for git to fix the line endings. A `git checkout -f` will do that though it will overwrite any changes you made. Caution is advised.

**Comment by mattkoch614** [comment]
Thanks @pbchase. We gave that a shot and it didn't appear to help out. What did end up getting us past this problem was the following suggestion: https://github.com/puphpet/puphpet/issues/266#issuecomment-193811336 which says we should run a ``dos2unix`` command:

```
dos2unix *.sh
dos2unix example.env.txt
```
This got rid of the ``==> default: /vagrant/.env: line 11: $'\r': command not found`` error.

The next error we see is the following, which appears to deal with ``netselect``:

```
==> default: Preparing to unpack .../netselect_0.3.ds1-26_amd64.deb ...
==> default: Unpacking netselect (0.3.ds1-26) ...
==> default: Selecting previously unselected package netselect-apt.
==> default: Preparing to unpack .../netselect-apt_0.3.ds1-26_all.deb ...
==> default: Unpacking netselect-apt (0.3.ds1-26) ...
==> default: Processing triggers for man-db (2.7.0.2-5) ...
==> default: Setting up netselect (0.3.ds1-26) ...
==> default: Setting up netselect-apt (0.3.ds1-26) ...
The SSH command responded with a non-zero exit status. Vagrant
assumes that this means the command failed. The output for this command
should be in the log above. Please read the output to determine what
went wrong.

```

**Comment by pbchase** [comment]
@mattkoch614, you are right that the dos2unix command will fix the issue, but only for you, this once. We need way to ensure the files are checked out correctly by the next person to clone this repo on Windows. I think my solution is right, but the testing procedure I gave you is poor. A better testing procedure might be 

1. Commit the .git_attributes file as previously described
2. Delete all files from the repo save for the .git folder
3. `git checkout master`
4. Run the command `file .env`.  It should respond

    $ file .env
    .env: ASCII text

For the net-select issues, I would need to see the output of your net-select command. You'll need to the ssh into the VM, become root and run the command.  The normal interaction looks like this:

```
→ vagrant ssh 

The programs included with the Debian GNU/Linux system are free software;
the exact distribution terms for each program are described in the
individual files in /usr/share/doc/*/copyright.

Debian GNU/Linux comes with ABSOLUTELY NO WARRANTY, to the extent
permitted by applicable law.
Last login: Tue Jan 30 09:59:55 2018 from 10.0.2.2
vagrant@redcap-deployment:~$ sudo su -
root@redcap-deployment:~# netselect-apt -c US
Using distribution stable.
Retrieving the list of mirrors from www.debian.org...

--2018-01-30 10:06:43--  https://www.debian.org/mirror/mirrors_full
Resolving www.debian.org (www.debian.org)... 128.31.0.62, 149.20.4.15, 2001:4f8:1:c::15
Connecting to www.debian.org (www.debian.org)|128.31.0.62|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 193336 (189K) [text/html]
Saving to: ‘/tmp/netselect-apt.bUBx5D’

/tmp/netselect-apt.bUBx5D           100%[====================================================================>] 188.80K  --.-KB/s   in 0.1s   

2018-01-30 10:06:43 (1.46 MB/s) - ‘/tmp/netselect-apt.bUBx5D’ saved [193336/193336]

Choosing a main Debian mirror using netselect.
(will filter only for mirrors in country US)
netselect: 7 (7 active) nameserver request(s)...         
Duplicate address 128.61.240.89 (http://debian.gtisc.gatech.edu/debian/, http://128.61.240.89/debian/); keeping only under first name.
netselect: 6 (6 active) nameserver request(s)...
Duplicate address 208.80.154.15 (http://mirrors.wikimedia.org/debian/, http://208.80.154.15/debian/); keeping only under first name.
netselect: 6 (6 active) nameserver request(s)...
Duplicate address 128.30.2.26 (http://debian.csail.mit.edu/debian/, http://128.30.2.26/debian/); keeping only under first name.
Running netselect to choose 10 out of 48 addresses.     
..................................................................................................................................................................................................................................................................................................................................................................................................
The fastest 10 servers seem to be:

	http://debian.gtisc.gatech.edu/debian/
	http://debian.osuosl.org/debian/
	http://debian.osuosl.org/debian/
	http://mirrors.wikimedia.org/debian/
	http://mirror.us.leaseweb.net/debian/
	http://mirror.cc.columbia.edu/debian/
	http://mirror.cogentco.com/debian/
	http://debian.mirror.constant.com/debian/
	http://mirror.yellowfiber.net/debian/
	http://debian.csail.mit.edu/debian/

Of the hosts tested we choose the fastest valid for HTTP:
        http://debian.gtisc.gatech.edu/debian/

Writing sources.list.
Done.
```

If you can show me the output of your `netselect-apt -c US` command run as root inside the VM I can give you better guidance.

**Comment by mattkoch614** [comment]
@pbchase Thanks, let me give this a shot!

By the way, cloning the most recent version of this repo now results in the following error:

```

**Comment by mattkoch614** [comment]
Thanks again for the testing procedures. I gave it a shot, but because the repo does not actually contain a ``.env`` file, I assume the ``.gitattributes`` file cannot process it when I perform a checkout. Instead, I added another line to the ``.gitattributes`` file, with the result being:

``*.sh crlf=input``
``.env crlf=input``
``*.txt crlf=input``

This way, the text file ``example.env.txt`` will also get processed by .gitattributes. After doing this, I got the output you described when performing a checkout, renaming ``example.env.txt`` to ``.env``, and then running ``file .env``:
```
$ file .env
.env: ASCII text
```

Moving on to the ``net-select`` issues - I'll reply to this thread once I'm able to grab output.

**Comment by pbchase** [comment]
@mattkoch614, I fixed the `deploy/plugins/autonotify3/99-<instance_name>-autonotify3.conf.example` bug. Sorry, I did that code review and let that pass.

As to the .gitattributes file that one needs to get added to the repo. Are you in a position to commit changes to a git repo? I don't know your background. If you've used git and github you'd certainly find it easy to send us a pull request, but if you have not dipped your toe into that pond there is a learning curve. I am happy to make the commit to add the .gitattributes file, but I don't want to steal your thunder if you want to make a contribution.

As to netselect, you got the result I expected. By default, the VM comes up with the address 192.168.33.113.  You'll see that in your .env file on the line `VM_IP=192.168.33.113`. That address must not be routable on your network. We chose that address because it is private address space, but it is routable on our network.  You'll need to find a suitable address that is routable on your network. You'll need to talk to your IT staff. You'll likely need to talk to a network engineer. 

I'm sorry. This is the downside of a design decision we had about a year back. There are good reasons to do what we did, but there are some negative consequences. Try your luck. If you fail, there is a plan B wherein we move the VM to a NAT within your computer.

**Comment by mattkoch614** [comment]
@pbchase  - Thanks! I have to say that you've been super helpful and thorough with all of your comments thus far. I think everything we are discussing here is a move in a positive direction. It seems like we are inching towards getting this in a workable state on Windows!

I'm happy to submit a PR - I'd love to make a contribution here :)

**Comment by pbchase** [comment]
@mattkoch614, I'm glad I can help and thrilled you can contribute.

**Comment by mattkoch614** [comment]
@pbchase - PR submitted!

**Comment by mattkoch614** [comment]
@pbchase Is there a way I can simply manually select a mirror instead of having ``netselect-apt`` try to find one for me? I'm not exactly sure how to modify the ``bootstrap.sh`` file to do this - I confirmed that ICMP is being blocked.

**Comment by pbchase** [comment]
@mattkoch614 This is definitely one where you pick your poison. Before we started using netselect-apt we had hardcoded repo mirrors.  This works fine until the get slow or go offline. I think what we'd want to do is add in a test of net-select apt.  If it fails, then we use a hardcoded mirror. If we can run net-select with a short timeout all the better.  

I'm not sure how to do that.  For all I know it is trivial, but I'm too busy to research it at the moment.  Could you do me a favor?  Could you open this specific item as its own issue? I'll get to it, but it'll be a day or two.

Thanks,
Philip

**Comment by mattkoch614** [comment]
No problem - I'll add it. I think perhaps we can close this one out, as the original title is somewhat unrelated at this point. Feel free to do this, or I can if you'd like.

**Comment by pbchase** [comment]
Thanks, @mattkoch614.
